# sourced by "drug_server.R"


# Drug welcome info -------------------------------------------------------

fn_drug_welcome <- function(){
  column(
    width = 12, offset = 0,
    shiny::tags$h1(
      class = "text-success text-left",
      shiny::icon(name = "angle-double-right", class = "fa-fw"), 
      "Drug Sensitivity Analysis"
    ),
    
    shiny::hr(),
    
    shiny::tags$p(
      class = "lead text-justify",
      "Genomic aberrations influence clinical responses to treatment and are potential biomarkers for drug screening. Drug sensitivity and gene expression profiling data of cancer cell lines in",
      shiny::tags$a("GDSC", href = "http://www.cancerrxgene.org/", target = "_blank"),
      " and ",
      shiny::tags$a("CTRP", href = "https://portals.broadinstitute.org/ctrp/", target = "_blank"),
      "are integrated for investigation. The expression of each gene in the gene set was performed by Spearman correlation analysis with the small molecule/drug sensitivity (IC50)."
    )
  )
}


fn_drug_help <- function(){
  column(
    width = 12, offset = 0,
    
    shiny::tags$div(
      class = "panel panel-primary",
      
      # head
      shiny::tags$div(
        class = "panel-heading",
        shiny::tags$h3(
          class = "panel-title text-left",
          shiny::tags$a(
            "data-toggle" = "collapse", "href" = "#help_drug",
            shiny::icon(name = "info-circle", class = "fa-fw"),
            "Click here for the detailed description of methods and results"
          )
        )
      ),
      
      # body
      shiny::tags$div(
        id = "help_drug", class = "panel-collapse collapse",
        shiny::tags$div(
          class = "panel-body",
          column(
            width = 12, offset = 0,
            # Methods
            shiny::tags$div(
              class = "bs-callout bs-callout-primary",
              shiny::tags$h3("Method"),
              
              shiny::tags$dl(
                class = "dl-horizontal",
                
                shiny::tags$dt("Data & Scripts:"),
                shiny::tags$dd(
                  "We collected 481 small molecules from", 
                  shiny::tags$a("Therapeutics Response Portal (CTRP)", href = "https://portals.broadinstitute.org/ctrp/", target = "_blank"), 
                  "and 265 small molecules from", 
                  shiny::tags$a("Genomics of Drug Sensitivity in Cancer (GDSC).", href = "http://www.cancerrxgene.org/", target = "_blank"),
                  "The whole analysis scripts used in the GSCALite were put on the",
                  shiny::tags$a("href" = "https://github.com/GuoBioinfoLab/GSCALite", shiny::icon(name = "github", lib = "font-awesome")), "."
                ),
                
                shiny::tags$dt("Drug Sensitivity: "),
                shiny::tags$dd(
                  "To analyze the correlation of gene expression and drug sensitivity, we downloaded the area under the dose-response curve (AUC) values for drugs and gene expression profiles for all cancer cell lines.",
                  "The method was used in", shiny::tags$a("Rees et al, 2016.", href = "https://www.nature.com/articles/nchembio.1986", target = "_blank"),
                  "The Pearson correlation coefficients of transcript levels and AUCs were normalized using Fisherâ€™s Z transformation. A Bonferroni-corrected, two-tailed distribution with family-wise error rate less than 0.025 in each tail for z-scored. Pearson correlation coefficients of annotated drug-target pairs were compared to the same number of correlation pairs generated by randomly sampling correlations."
                )
              )
            ),
            
            # Results
            shiny::tags$div(
              class = "bs-callout bs-callout-danger",
              shiny::tags$h3("Result"),
              
              shiny::tags$dl(
                class = "dl-horizontal",
                
                shiny::tags$dt("Drug Sensitivity: "),
                shiny::tags$dd(
                  "The gene set drug resistence analysis from", 
                  shiny::tags$a("GDSC", href = "http://www.cancerrxgene.org/", target = "_blank"),
                  "/",
                  shiny::tags$a("CTRP", href = "https://portals.broadinstitute.org/ctrp/", target = "_blank"), 
                  "IC50 drug data. The Spearman correlation represent the the gene expression correlates with the drug. The positive correlation means that the gene high expression is resistant to the drug, vise verse.")
              )
            )
          )
        )
      )
    )
  )
}

# Drug output -------------------------------------------------------------
drugOutput <- function(id) {
  ns <- NS(id)
  column(
    width = 12, offset = 0,
    shinydashboard::tabBox(
      id = "expr_plot", title = "", width = 12,
      # drug output
      tabPanel(
        title = "GDSC",
        uiOutput(ns("gdsc_massage")),
        column(width=2,download_bt(ns("gdsc"))
        ),
        column(width = 12,
               plotOutput(outputId = ns("gdsc"), height = "100%") %>% withSpinner(color = "#0dc5c1",size = 0.5, proxy.height = "200px")
               )
      ),
      tabPanel(
        title = "CTRP",
        uiOutput(ns("ctrp_massage")),
        column(width=2,
               download_bt(ns("ctrp"))
        ),
        column(width = 12,
        plotOutput(outputId = ns("ctrp"), height = "100%") %>% withSpinner(color = "#0dc5c1",size = 0.5, proxy.height = "200px")
        )
      )
    )
  )
}

fn_drug_result <- function(.test){
  if (.test == TRUE) {
    drugOutput("drug")
  } else{
    column(
      width = 12, offset = 0,
      shiny::tags$div(style = "height=500px;", class = "jumbotron", shiny::tags$h2("This analysis is not selected"))
    )
  }
}

# drug analysis -----------------------------------------------------------


fn_filter_drug <- function(.x, .cor = 0.2, .fdr = 0.05) {
  .x %>% dplyr::filter(abs(cor_sprm) > .cor, fdr < .fdr)
}
fn_filter_drug_ctrp <- function(.x, .cor = 0.2, .fdr = 0.05) {
  .x %>% dplyr::filter(abs(cor_sprm) > .cor, p_val < .fdr)
}
# GDSC --------------------------------------------------------------------


gdsc_plot <- function(gdsc_gene_list_sig_drug,t_gdsc) {
  print(glue::glue("{paste0(rep('-', 10), collapse = '')} Start GDSC Plot @ {Sys.time()} {paste0(rep('-', 10), collapse = '')}"))
  print(head(gdsc_gene_list_sig_drug))
    gdsc_gene_list_sig_drug %>%
      dplyr::mutate(
        fdr = ifelse(-log10(fdr) > 40, 40, -log10(fdr)),
        cor_sprm = ifelse(cor_sprm > 0.4, 0.4, cor_sprm),
        cor_sprm = ifelse(cor_sprm < -0.4, -0.4, cor_sprm)
      ) %>%
      dplyr::left_join(t_gdsc, by = "drug_name") -> gdsc_plot_ready
    
    
    gdsc_plot_ready %>%
      dplyr::group_by(symbol) %>%
      dplyr::summarise(cor_sum = sum(cor_sprm)) %>%
      dplyr::arrange(cor_sum) -> gdsc_gene_rank
    
    gdsc_plot_ready %>%
      dplyr::distinct(drug_name, target_pathway, count) %>%
      dplyr::group_by(target_pathway) %>%
      dplyr::mutate(per = dplyr::n() / count) %>%
      dplyr::arrange(per) %>%
      dplyr::ungroup() %>%
      dplyr::select(drug_name, target_pathway, per) -> gdsc_drug_per
    
    gdsc_plot_ready %>%
      dplyr::left_join(gdsc_drug_per, by = c("drug_name", "target_pathway")) %>%
      dplyr::group_by(drug_name) %>%
      dplyr::mutate(drug_count = dplyr::n()) %>%
      dplyr::ungroup() %>%
      dplyr::arrange(per, target_pathway, drug_count) %>%
      dplyr::select(drug_name, target_pathway, drug_count, per) %>%
      dplyr::distinct() %>%
      dplyr::mutate(target_pathway = stringr::str_to_title(target_pathway)) -> gdsc_drug_rank_pre
    
    gdsc_drug_rank_pre %>%
      dplyr::distinct(target_pathway, per) %>%
      dplyr::arrange(per, target_pathway) -> .foo
    
    pathway_color <-
      .foo %>%
      dplyr::mutate(color = ggthemes::gdocs_pal()(nrow(.foo)))
    
    gdsc_drug_rank_pre %>%
      dplyr::select(-per) %>%
      dplyr::left_join(pathway_color, by = "target_pathway") -> drug_rank
    
    p <-
      gdsc_plot_ready %>%
      ggplot(aes(x = symbol, y = drug_name, color = cor_sprm)) +
      geom_point(aes(size = fdr)) +
      scale_x_discrete(limits = gdsc_gene_rank$symbol, expand = c(0.012, 0.012)) +
      scale_y_discrete(limits = drug_rank$drug_name, expand = c(0.012, 0.012), position = "right") +
      scale_color_gradient2(
        name = "Spearman Correlation",
        high = "red",
        mid = "white",
        low = "blue"
      ) +
      scale_size_continuous(
        name = "-log10(FDR)"
      ) +
      # ggthemes::theme_gdocs() +
      theme(
        panel.background = element_rect(color = "black", fill = "white", size = 0.1),
        panel.grid = element_line(colour = "grey", linetype = "dashed"),
        panel.grid.major = element_line(colour = "grey", linetype = "dashed", size = 0.2),
        
        axis.title = element_blank(),
        axis.text.x = element_text(size = 9, angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(size = 10, color = drug_rank$color),
        
        axis.ticks = element_line(color = "black"),
        
        legend.position = "bottom",
        legend.direction = "horizontal",
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 10),
        # legend.key.width = unit(1,"cm"),
        # legend.key.heigh = unit(0.3,"cm"),
        legend.key = element_rect(fill = "white", colour = "black")
      ) + guides(
        color = guide_colorbar(
          title.position = "top",
          title.hjust = 0.5,
          barheight = 0.5,
          barwidth = 10
        )
      )
    print(glue::glue("{paste0(rep('-', 10), collapse = '')} End GDSC Plot @ {Sys.time()} {paste0(rep('-', 10), collapse = '')}"))
    p
}

# CTRP --------------------------------------------------------------------


ctrp_plot <- function(ctrp_gene_list_sig_drug) {
  print(glue::glue("{paste0(rep('-', 10), collapse = '')} Start CTRP plot @ {Sys.time()} {paste0(rep('-', 10), collapse = '')}"))
    print(head(ctrp_gene_list_sig_drug))
    ctrp_gene_list_sig_drug %>%
      dplyr::mutate(
        p_val = ifelse(-log10(p_val) > 50, 50, -log10(p_val)),
        cor_sprm = ifelse(cor_sprm > 0.5, 0.5, cor_sprm),
        cor_sprm = ifelse(cor_sprm < -0.5, -0.5, cor_sprm)
      ) -> ctrp_plot_ready
    
    ctrp_plot_ready %>%
      dplyr::group_by(symbol) %>%
      dplyr::summarise(cor_sum = sum(cor_sprm)) %>%
      dplyr::arrange(cor_sum) -> ctrp_gene_rank
    
    p <- ctrp_plot_ready %>%
      ggplot(aes(x = symbol, y = drug_name, color = cor_sprm)) +
      geom_point(aes(size = p_val)) +
      scale_x_discrete(limits = ctrp_gene_rank$symbol, expand = c(0.012, 0.012)) +
      scale_y_discrete(expand = c(0.012, 0.012), position = "right") +
      scale_color_gradient2(
        name = "Spearman Correlation",
        high = "red",
        mid = "white",
        low = "blue"
      ) +
      scale_size_continuous(
        name = "-log10(FDR)"
      ) +
      # ggthemes::theme_gdocs() +
      theme(
        panel.background = element_rect(color = "black", fill = "white", size = 0.1),
        panel.grid = element_line(colour = "grey", linetype = "dashed"),
        panel.grid.major = element_line(colour = "grey", linetype = "dashed", size = 0.2),
        
        axis.title = element_blank(),
        axis.text.x = element_text(
          size = 9,
          angle = 90,
          hjust = 1,
          vjust = 0.5
        ),
        axis.text.y = element_text(
          # color = drug_rank$color,
          size = 10
        ),
        
        axis.ticks = element_line(color = "black"),
        
        legend.position = "bottom",
        legend.direction = "horizontal",
        legend.text = element_text(size = 10),
        legend.title = element_text(size = 10),
        # legend.key.width = unit(1,"cm"),
        # legend.key.heigh = unit(0.3,"cm"),
        legend.key = element_rect(fill = "white", colour = "black")
      ) + guides(
        color = guide_colorbar(
          title.position = "top",
          title.hjust = 0.5,
          barheight = 0.5,
          barwidth = 10
        )
      )
    print(glue::glue("{paste0(rep('-', 10), collapse = '')} End CTRP plot @ {Sys.time()} {paste0(rep('-', 10), collapse = '')}"))
    return(p)
}
