# sourced by 'tcga_mirna_ui.R'
# save as 'tcga_mirna_help.R'
# help file for tcga_mirna ui part

fluidRow(
  style = "width:85%",
  column(
    width = 12,
    # collapsable panel
    shiny::tags$div(
      class = "panel panel-primary",
      
      shiny::tags$div(
        class = "panel-heading",
        shiny::tags$h3(
          class = "panel-title text-left",
          shiny::tags$a(
            "data-toggle" = "collapse", "href" = "#help_tcga_mirna",
            shiny::icon(name = "info-circle"),
            "Click here for the detailed description of methods and results"
          )
        )
      ),
      
      shiny::tags$div(
        id = "help_tcga_mirna", class = "panel-collapse collapse",
        shiny::tags$div(
          class = "panel-body",
          # here comes the content
          ###############
          column(
            width = 12,
            # overall description----
            shiny::div(
              class = "bs-callout bs-callout-primary",
              shiny::tags$h3("Method"),
              shiny::tags$dl(
                class="dl-horizontal",
                shiny::tags$dt("Data"),
                shiny::tags$dd(
                  shiny::tags$p("miRNA regulation data was collected form databases: include experimental verified (papers, ",
                  shiny::tags$a(href = "http://diana.imis.athena-innovation.gr/DianaTools/index.php?r=site/index", 
                                shiny::tags$span("TarBase, "),target = "_blank"), 
                  shiny::tags$a(href = "http://mirtarbase.mbc.nctu.edu.tw/php/index.php", 
                                shiny::tags$span("miRTarBase, "),target = "_blank"),
                  shiny::tags$a(href = "http://www.mir2disease.org/", 
                                shiny::tags$span("mir2disease"),target = "_blank"),"), and ",
                  shiny::tags$a(href = "http://www.targetscan.org/vert_70/", 
                                shiny::tags$span("targetscan, "),target = "_blank"),
                  shiny::tags$a(href = "http://34.236.212.39/microrna/home.do", 
                                shiny::tags$span("miRanda"),target = "_blank"), 
                  " predicted. And only miRNA-gene pairs who have been recorded will be used to calculate a expression correlation here."),
                  shiny::tags$p(
                    "miRNA transcript expression data was collected from TCGA, including 9105 samples and 33 cancer types."
                  )
                ),
                
                shiny::tags$dt("Regulation confirm"),
                shiny::tags$dd(
                  shiny::tags$p("1. miRNA expression and gene expression were merged by TCGA barcode. We test the association between paired mRNA  and miRNA expression, based on Person's product moment correlation coefficient, and follows a t distribution. P-value was adjusted by FDR and genes with FDR<=0.05, R<0 will be remained. Correlation was calculated in all paired samples(33 cancers), so cancer type selection change will not change the result with no change in gene set.",
                                "2. In consideration of the presence of positive regulators like transcription factors, a miRNA-gene pair with negative correlation will be considered as a potential negatively regulation pair."),
                  
                  shiny::tags$p("3. Only the miRNA-gene pairs who have been recorded in databases we refered below will be calculated at step 1.")
                ),
                
                shiny::tags$dt(shiny::tags$p("Network"), 
                               shiny::tags$p("construction")),
                shiny::tags$dd(
                  shiny::tags$p("Network was generated by R packages: ",
                                shiny::tags$a(href = "https://christophergandrud.github.io/networkD3/", 
                                              shiny::tags$span("networkD3 "),target = "_blank"),
                                "and",
                                shiny::tags$a(href = " https://datastorm-open.github.io/visNetwork/",
                                              shiny::tags$span("visNetwork. "),target = "_blank"),
                                "Subgroups of networkD3 is generated by getting communities in a graph via random walks in igraph R package.")
                )
              )
            ),
            
            shiny::div(
              class = "bs-callout bs-callout-danger",
              shiny::tags$h3("Results"),
              shiny::tags$dl(
                class="dl-horizontal",
                shiny::tags$dt("networkD3"),
                shiny::tags$dd(
                  "The networkD3 R package is used to generate this beautiful network, if you need to get a cool miRNA regulation network for your paper, it will be a nice choice. In this network, node size is positively correlate to the node's degree. Subgroups are generated by getting communities in a graph via random walks by igraph R package, and shown by different colors."
                ),
                
                shiny::tags$dt("visNetwork"),
                shiny::tags$dd(
                  "The visNetwork R package is used to generate this useful network, you can click, drag, delele and add nodes or edges on the web as you wish. A node represents a miRNA or gene, an edge represents a regulation of miRNA to gene. We have cluster them by color, and Node size is positively correlate to the node's degree similar to networkD3, and edge width is defined by absolute value of correlation coefficient."
                )
              )
            )
          ) # all description end
        )
      )
    )
  )
)
