# sourced by 'tcga_snv_ui.R'
# save as 'tcga_snv_help.R'
# help file for tcga_snv ui part


fluidRow(
  style = "width:85%",
  column(
    width = 12,
    # collapsable panel
    shiny::tags$div(
      class = "panel panel-primary",
      shiny::tags$div(
        class = "panel-heading",
        shiny::tags$h3(
          class = "panel-title text-left",
          shiny::tags$a(
            "data-toggle" = "collapse", "href" = "#help_tcga_snv",
            shiny::icon(name = "info-circle"),
            "Click here for the detailed description of methods and results"
          )
        )
      ),
      shiny::tags$div(
        id = "help_tcga_snv", class = "panel-collapse collapse",
        shiny::tags$div(
          class = "panel-body",
          # here comes the content
          ###############
          column(
            width = 12,

            shiny::tags$div(
              class = "bs-callout bs-callout-primary",
              shiny::tags$h3("Method"),

              shiny::tags$dl(
                class = "dl-horizontal",
                shiny::tags$dt("Data"),
                shiny::tags$dd(
                  "We collected 8663 SNV data from",
                  shiny::tags$a("href" = "https://gdc.cancer.gov/", "NCI Genomic Data Commons,"),
                  "including 33 cancer types. The TCGA SNV data include variant type values: Missense_Mutation, Silent, 5'Flank, 3'UTR, RNA, In_Frame_Del, Nonsense_Mutation, Splice_Site, Intron, 5'UTR, In_Frame_Ins, Frame_Shift_Del, Nonstop_Mutation, 3'Flank, Frame_Shift_Ins, Translation_Start_Site. GSCALite filters out Silent, Intron, IGR, 3'UTR, 5'UTR, 3'Flank and 5'Flank for SNV percentage calculation."
                ),

                shiny::tags$dt("SNV percentage"),
                shiny::tags$dd(
                  "SNV percentage of each gene's coding region was calculate by: Num Of Mutated Sample/Num of Cancer Sample. The SNV summary, oncoplot and survival were all based on this data."
                ),

                shiny::tags$dt(shiny::tags$p("SNV "),
                               shiny::tags$p("summary/oncoplot")),
                shiny::tags$dd(
                  "SNV summary and oncoplot waterfall plot was generated by maftools ",
                  shiny::tags$a(href = "https://www.biorxiv.org/content/early/2016/05/11/052662", "(Mayakonda and Koeffler, 2016)", target = "_blank")
                ),

                shiny::tags$dt("SNV survival"),
                shiny::tags$dd(
                  "SNV data and clinical overall survival data was combined, and R package ",
                  shiny::tags$a(href="https://cran.r-project.org/web/packages/survival/index.html","survival ", target = "_blank"),
                  "was used to estimate survival difference between mutate and non-mutate gene. Cox regression ",
                  shiny::tags$a(href="https://www.jstor.org/stable/2240714?seq=1#page_scan_tab_contents","(Andersen, P. and Gill, R, 1982.)", target = "_blank"),
                  "was performed to estimate the hazards of mutated group, and a log rank test ",
                  shiny::tags$a(href="https://academic.oup.com/biomet/article-abstract/69/3/553/221086?redirectedFrom=fulltext", "(Harrington, D. P. and Fleming, T. R.)", target = "_blank"),
                  " was also performed to compare the distributions of two groups, p value <=0.05 was considered as significant."
                )
              )
            ),

            shiny::tags$div(
              class = "bs-callout bs-callout-danger",
              shiny::tags$h3("Results"),

              shiny::tags$dl(
                class = "dl-horizontal",
                shiny::tags$dt("SNV percentage"),
                shiny::tags$dd(
                  "Give you SNV frequency of genes in each cancers. The deeper of color, the higher of mutate frequency. Numbers in each cells represent number of samples have corresponding mutated gene in corresponding cancers. The 0 in the cell indicates there is no mutation in the gene coding region, and blank in the cell indicates there is no mutation in all regions of a gene."
                ),

                shiny::tags$dt("SNV summary"),
                shiny::tags$dd(
                  "A summary plot displays number of variants in each sample as a stacked barplot and variant types as a boxplot summarized."
                ),

                shiny::tags$dt("SNV oncoplot"),
                shiny::tags$dd(
                  "An oncoplot also known as waterfall plots, gives a mutation distribution of top 10 (when num of gene set <10, show all) mutated genes in your gene set and a SNV classification of SNV types (include missense mutation, frame shift deletion, nonsense mutation etc.). All selected cancers' sample will be shown together. Side barplot and top barplots show number of variants in each sample or each gene."
                ),

                shiny::tags$dt("SNV survival"),
                shiny::tags$dd(
                  "A survival plot will give you a survival difference between mutated and non-mutate gene, only p value significantly(<=0.05) genes will be displayed here.",
                  shiny::tags$span("Blue points", style = "color:blue"),
                  " represent patients with mutated gene have worse overall survival, ",
                  shiny::tags$span("red points", style = "color:red"),
                  "represent patients with mutated gene have better overall survival."
                )
              )
            )
          )
        )
      )
    )
  )
)